src/rgw/rgw_data_sync.cc

```diff
@@ -832,94 +832,125 @@ struct bucket_instance_meta_info {
+	sync_env->svc->zone->get_zone_params().log_pool,
+	oid_prefix);
       yield; // yield so OmapAppendCRs can start
 
       do {
         yield {
-          string entrypoint = string("/admin/metadata/bucket.instance");
+          string entrypoint = "/admin/metadata/bucket.instance"s;
 
           rgw_http_param_pair pairs[] = {{"max-entries", "1000"},
                                          {"marker", result.marker.c_str()},
                                          {NULL, NULL}};
 
-          call(new RGWReadRESTResourceCR<read_metadata_list>(sync_env->cct, sc->conn, sync_env->http_manager,
-                                                             entrypoint, pairs, &result));
-        }
-        if (retcode < 0) {
-          ldpp_dout(dpp, 0) << "ERROR: failed to fetch metadata for section bucket.instance" << dendl;
+          call(new RGWReadRESTResourceCR<read_metadata_list>(
+		 sync_env->cct, sc->conn, sync_env->http_manager,
+		 entrypoint, pairs, &result));
+	}
+	if (retcode < 0) {
+	  ldpp_dout(dpp, 0)
+	    << "ERROR: failed to fetch metadata for section bucket.instance"
+	    << dendl;
           return set_cr_error(retcode);
         }
 
         for (iter = result.keys.begin(); iter != result.keys.end(); ++iter) {
-          ldpp_dout(dpp, 20) << "list metadata: section=bucket.instance key=" << *iter << dendl;
+          ldpp_dout(dpp, 20) << "list metadata: section=bucket.instance key="
+			     << *iter << dendl;
           key = *iter;
 
           yield {
             rgw_http_param_pair pairs[] = {{"key", key.c_str()},
                                            {NULL, NULL}};
 
-            call(new RGWReadRESTResourceCR<bucket_instance_meta_info>(sync_env->cct, sc->conn, sync_env->http_manager, path, pairs, &meta_info));
-          }
-
-          num_shards = meta_info.data.get_bucket_info().layout.current_index.layout.normal.num_shards;
-          if (num_shards > 0) {
-            for (i = 0; i < num_shards; i++) {
-              char buf[16];
-              snprintf(buf, sizeof(buf), ":%d", i);
-              s = key + buf;
-              yield entries_index->append(s, sync_env->svc->datalog_rados->get_log_shard_id(meta_info.data.get_bucket_info().bucket, i));
-            }
-          } else {
-            yield entries_index->append(key, sync_env->svc->datalog_rados->get_log_shard_id(meta_info.data.get_bucket_info().bucket, -1));
+            call(new RGWReadRESTResourceCR<bucket_instance_meta_info>(
+		   sync_env->cct, sc->conn, sync_env->http_manager, path, pairs,
+		   &meta_info));
           }
-        }
-        truncated = result.truncated;
+	  if (retcode < 0) {
+	    ldpp_dout(dpp, 0) << "ERROR: failed to fetch metadata for key: "
+			      << key << dendl;
+	    return set_cr_error(retcode);
+	  }
+	  yield entries_index->append(
+	    fmt::format("{}:{}", key, 0),
+	    sync_env->svc->datalog_rados->get_log_shard_id(
+	      meta_info.data.get_bucket_info().bucket, 0));
```

Comment:
i think a comment would help here. something like:
```c++
// now that bucket full sync is bucket-wide instead of per-shard, we only need to register a single shard
// of each bucket to guarantee that sync will see everything that happened before data full sync starts.
// this also means we don't have to care about the bucket's current shard count
```